### DISCUSSION TITLE
Should CGAL::Polygon_mesh_processing::refine() remove vertices?

### DISCUSSION THREAD

**Discussion by pentacular** [CGAL/cgal#7948] [Q&A]
I expected that [CGAL::Polygon_mesh_processing::refine()](https://doc.cgal.org/latest/Polygon_mesh_processing/group__PMP__meshing__grp.html#gafd90c169a7c17a1e503651f59a6dc57b) would just add new vertices, but I see what looks like vertex removal happening as well.

The below is expected to be a cube.

![image](https://github.com/CGAL/cgal/assets/33278307/609565cd-3b6d-4b73-be61-65fc96c7f92c)

Is this expected behavior?

If not, I'll put together a minimal test case.

**Comment by sloriot** [comment]
I think there is no removal, but the placement of new vertices might be surprising.

**Comment by pentacular** [comment]
Ok, I have a pretty minimal test case.

https://gist.github.com/pentacular/2bc66d9c01439e70a32cc083a4ec61d5

It takes in a rectangular prism

![image](https://github.com/CGAL/cgal/assets/33278307/d25802f0-229c-483c-a4bc-f119dcdbb5b9)

And refines it to something that isn't. :)

![image](https://github.com/CGAL/cgal/assets/33278307/45784498-98dd-489a-b374-9880f53c3ee8)

Can you see what I am doing wrong?

**Comment by pentacular** [comment]
Ok, I had a chance to dig into this a bit deeper.

The issue is due to non-coplanar edge-flips introducing edges that cut through existing corners.

Once I add a restriction that the edge-flips be between coplanar triangles the shape is preserved as I had expected.

Is my expectation that refine should not change the shape correct, or have I misunderstood refinement?

**Reply by pentacular** [reply]
@sloriot any thoughts?
